### ISSUE TITLE
many2many, Preload, and Query callbacks

### ISSUE THREAD

**Issue by glb** [jinzhu/gorm#1621]
### What version of Go are you using (`go version`)?
go version go1.8.3 darwin/amd64

### Which database and its version are you using?
sqlite3

### What did you do?
I would like to implement a transparent field encryption plugin for `jinzhu/gorm` that lets the developer tag their fields as needing encryption and the plugin does the required work to transparently encrypt / decrypt the data that is stored in the database.

I have implemented two callback functions that encrypt / decrypt the field, and they are getting called properly when I save / create objects as well as when I use `Find` / `First` / ... to retrieve a single level of object.

However, when I have a `many2many` association and try to use `Preload` to load the associated objects, the `Query` callback is not called. Instead, the `RowQuery` callback gets called, but unfortunately (I believe) this callback is not able to do what I need it to do. For other association types (`belongs-to`, `has-one`, `has-many`), the `Query` callback is called and is able to do the right thing.

*Original proposal:* I am currently thinking that it would be _very_ helpful to have a new callback chain that would be called for each preloaded object after it has been scanned, like:
```go
scope.scan(rows, columns, append(fields, joinTableFields...)) // <-- this already exists
scope.New(elem.Addr().Interface()).callCallbacks(scope.db.parent.callbacks.preloads) // <-- proposal
```

I would prefer to be able to re-use the `Query` callback chain, but that includes the pre-registered `gorm:query` callback and I'm pretty sure that callback would not like to be called in this context.

If you agree with this proposal, I think I can implement it; I already have something that appears to work for `many2many` associations.

**Update**: I am no longer certain that this proposal is a good one since the only association type where the callbacks aren't working as desired with `Preload` is `many2many`.

### A short example of the problem
```go
package main

import (
	"fmt"

	"github.com/jinzhu/gorm"

	// _ "github.com/jinzhu/gorm/dialects/mssql"
	// _ "github.com/jinzhu/gorm/dialects/mysql"
	// _ "github.com/jinzhu/gorm/dialects/postgres"
	_ "github.com/jinzhu/gorm/dialects/sqlite"
)

type Account struct {
	ID         uint64
	Name       string
	Secret     string     `gorm:"encrypted"`
	Owners     []Owner    `gorm:"many2many:account_owners;AssociationForeignKey:ID;ForeignKey:ID"`
	CreditCard CreditCard // has-one
	Documents  []Document // has-many
}

type Toy struct {
	ID        uint64
	AccountID uint64
	Account   Account // belongs-to
	Secret    string  `gorm:"encrypted"`
}

type CreditCard struct {
	ID        uint64
	AccountID uint64
	Number    string `gorm:"encrypted"`
}

type Owner struct {
	ID       uint64
	Name     string
	Password string `gorm:"encrypted"`
}

type Document struct {
	ID        uint64
	AccountID uint64
	Name      string
	Content   string `gorm:"encrypted"`
}

// This sample doesn't actually encrypt / decrypt the field values, just replaces with
// "encrypted" or "decrypted" depending on the operation.

func encryptFieldCallback(scope *gorm.Scope) {
	for _, field := range scope.Fields() {
		if _, isEncrypted := field.TagSettings["ENCRYPTED"]; isEncrypted {
			// ... get and encrypt value
			scope.SetColumn(field.DBName, "encrypted")
		}
	}
}

func decryptFieldCallback(scope *gorm.Scope) {
	fmt.Printf("decryptFieldCallback called with scope.Value == %v %v\n", scope.IndirectValue().Type(), scope.Value)

	// ... get, decrypt, and store value
	var (
		isSlice, isPtr bool
		resultType     reflect.Type
		results        = scope.IndirectValue()
	)

	if kind := results.Kind(); kind == reflect.Slice {
		isSlice = true
		resultType = results.Type().Elem()
		if resultType.Kind() == reflect.Ptr {
			isPtr = true
			resultType = resultType.Elem()
		}
	}

	if isSlice {
		for i := 0; i < results.Len(); i++ {
			result := results.Index(i)

			if isPtr {
				result = result.Elem()
			}

			decryptStruct(scope, result)
		}
	} else if isPtr {
		decryptStruct(scope, results.Elem())
	} else {
		decryptStruct(scope, results)
	}
}

func decryptStruct(scope *gorm.Scope, result reflect.Value) {
	for _, x := range scope.Fields() {
		if _, ok := x.TagSettings["ENCRYPTED"]; ok {
			field := result.FieldByName(x.Name)
			decrypted := "decrypted"
			if field.Kind() == reflect.Ptr {
				field.Set(reflect.ValueOf(&decrypted))
			} else {
				field.Set(reflect.ValueOf(decrypted))
			}
		}
	}
}

func decryptRowCallback(scope *gorm.Scope) {
	// this gets called for many2many, but are you supposed to map the Rows into the objects?
	// does the cursor get rewound somehow after the gorm:row_query callback runs? Are you supposed
	fmt.Printf("decryptRowCallback called with scope.Value == %v %v\n", scope.IndirectValue().Type(), scope.Value)
}

var db *gorm.DB

func init() {
	var err error
	db, err = gorm.Open("sqlite3", "test.db")
	// Please use below username, password as your database's account for the script.
	// db, err = gorm.Open("postgres", "user=gorm dbname=gorm sslmode=disable")
	// db, err = gorm.Open("mysql", "gorm:gorm@/dbname?charset=utf8&parseTime=True")
	// db, err = gorm.Open("mssql", "sqlserver://gorm:LoremIpsum86@localhost:1433?database=gorm")
	if err != nil {
		panic(err)
	}

	db.LogMode(true)
}

func main() {
	db.Callback().Create().Before("gorm:create").Register("encrypt_fields", encryptFieldCallback)
	db.Callback().Query().After("gorm:query").Register("decrypt_fields", decryptFieldCallback)
	db.Callback().RowQuery().After("gorm:row_query").Register("decrypt_row", decryptRowCallback)

	db.AutoMigrate(&Account{}, &Owner{}, &CreditCard{}, &Document{}, &Toy{})

	account := Account{
		Name:   "account",
		Secret: "secret",
		Owners: []Owner{
			Owner{Name: "owner", Password: "password"},
		},
		CreditCard: CreditCard{Number: "1234"},
		Documents: []Document{
			Document{Name: "name", Content: "content"},
		},
	}

	db.Create(&account)

	toy := Toy{
		AccountID: account.ID,
		Account:   account,
		Secret:    "secret",
	}

	db.Create(&toy)

	createdAccount := Account{ID: account.ID}
	createdToy := Toy{ID: toy.ID}

	// using `Preload()`
	db.
		Preload("CreditCard"). // works (has-one)
		Preload("Owners").     // doesn't work (many2many)
		Preload("Documents").  // works (has-many)
		First(&createdAccount, &createdAccount)

	db.
		Preload("Account").
		First(&createdToy, &createdToy) // works (belongs-to)

	// using `Related()`
	// db.First(&createdAccount, &createdAccount)
	// db.Model(&createdAccount).Related(&createdAccount.CreditCard, "CreditCard") // works (has-one)
	// db.Model(&createdAccount).Related(&createdAccount.Owners, "Owners")         // works (many2many)
	// db.Model(&createdAccount).Related(&createdAccount.Documents, "Documents")   // works (has-many)

	// db.First(&createdToy, &createdToy)
	// db.Model(&createdToy).Related(&createdToy.Account, "Account") // works (belongs-to)

	// y, _ := json.Marshal(&createdAccount)
	// fmt.Print("account ")
	// os.Stdout.Write(y)
	// fmt.Println()

	// z, _ := json.Marshal(&createdToy)
	// fmt.Print("toy ")
	// os.Stdout.Write(z)
	// fmt.Println()

	failed := false

	if createdAccount.Secret == "encrypted" {
		fmt.Println("failed to decrypt account secret (struct)")
		failed = true
	}

	if createdAccount.CreditCard.Number == "encrypted" {
		fmt.Println("failed to decrypt credit card number (has-one)")
		failed = true
	}

	if createdAccount.Owners[0].Password == "encrypted" {
		fmt.Println("failed to decrypt owner password (many2many)")
		failed = true
	}

	if createdAccount.Documents[0].Content == "encrypted" {
		fmt.Println("failed to decrypt document content (has-many")
		failed = true
	}

	if createdToy.Account.Secret == "encrypted" {
		fmt.Println("failed to decrypt account secret (belongs-to)")
		failed = true
	}

	if !failed {
		fmt.Println("success")
	}
}
```

**Comment by glb** [comment]
I would label this as `type:feature` but I don't seem to have permission to do so.

**Comment by glb** [comment]
I'm working now on expanding the tests done in the sample application to characterize the behaviour of the other association types when using `Preload` versus using `Related`.

**Comment by glb** [comment]
I've done the characterization and it appears that the only issue is with trying to use `many2many` and `Preload`: the data gets loaded but the `Query` callback chain isn't called, only the `RowQuery` callback chain, and I'm not sure how to make the `RowQuery` callback chain do what needs to get done or even if it is possible to do so.

I am tempted to insert a mechanism to bypass the work done by `queryCallback` so that `handleManyToManyPreload` can call the  `QueryCallback` callback chain after it's scanned the row. I think this might make the function work as I'd expected it to without introducing the complexity of a new callback chain.

Thoughts?

**Comment by glb** [comment]
I've become more convinced that this is a bug rather than a feature request -- the `many2many` association appears to be the only one that doesn't call the `Query` callback chain with preloading. I've fixed the behaviour in #1622.

**Comment by afranioce** [comment]
I have this problem with many2many
